### PR TITLE
 Implement dohpath SvcParam

### PR DESCRIPTION
**Pull request by hanazuki**
_Depends on #32_

This patch implements "dohpath" SvcParam proposed in [draft-ietf-add-svcb-dns-08]. This parameter specifies a URI template for the :path used in DNS-over-HTTPS requests.

"dohpath" is employed by [DDR], also a to-be-published Proposed Standard that specifies how to upgrade DNS transport to a more secure one, i.d., DNS-over-TLS or DNS-over-HTTPS. DDR is deployed in the public DNS resolvers including Cloudflare DNS, Google Public DNS, and Quad9.

[draft-ietf-add-svcb-dns-08]: https://datatracker.ietf.org/doc/draft-ietf-add-svcb-dns/08/
[DDR]: https://datatracker.ietf.org/doc/draft-ietf-add-ddr/

### CODE REVIEW

**Comment by sorah** [comment]
The draft has been published as https://datatracker.ietf.org/doc/rfc9461/

**Review by sorah — APPROVED** [review]

**Comment by hsbt** [comment]
@sorah Could you handle this and https://github.com/ruby/resolv/pull/32?

**Comment by sorah** [comment]
🙆🏻‍♀️

**Comment by hanazuki** [comment]
RFC 9461 has no difference in the format of "dohpath" from the draft that the patch is originally based on (updated the patch).

File: lib/resolv.rb [hunk 1]
```diff
@@ -1676,6 +1684,379 @@ def get_rr
...
+          @template = template.encode('utf-8')
+        end
+
+        def encode(msg) # :nodoc:
+          msg.put_bytes(@template)
+        end
+
+        def self.decode(msg) # :nodoc:
+          template = msg.get_bytes.force_encoding('utf-8')
+          return self.new(template)
```
Comment (olleolleolle) [inline lib/resolv.rb:2054]: This would be syntactically the same, right?

```suggestion
          new(template)
```

File: lib/resolv.rb [hunk 2]
```diff
@@ -2339,6 +2720,84 @@ def self.decode_rdata(msg) # :nodoc:
...
+            msg.put_pack("n", @priority)
+            msg.put_name(@target, compress: false)
+            @params.encode(msg)
+          end
+
+          def self.decode_rdata(msg) # :nodoc:
+            priority, = msg.get_unpack("n")
+            target    = msg.get_name
+            params    = SvcParams.decode(msg)
+            return self.new(priority, target, params)
```
Comment (olleolleolle) [inline lib/resolv.rb:2780]: (Same as above.)

```suggestion
            new(priority, target, params)
```

File: lib/resolv.rb [hunk 3]
```diff
@@ -1676,6 +1684,379 @@ def get_rr
...
+
+        def encode(msg) # :nodoc:
+          @addresses.each do |address|
+            msg.put_bytes(address.address)
+          end
+        end
+
+        def self.decode(msg) # :nodoc:
+          addresses = msg.get_list { IPv6.new(msg.get_bytes(16)) }
+          return self.new(addresses)
```
Comment (olleolleolle) [inline lib/resolv.rb:2024]:
```suggestion
          new(addresses)
```

File: lib/resolv.rb [hunk 4]
```diff
@@ -1676,6 +1684,379 @@ def get_rr
...
+          @port = port.to_int
+        end
+
+        def encode(msg) # :nodoc:
+          msg.put_pack('n', @port)
+        end
+
+        def self.decode(msg) # :nodoc:
+          port, = msg.get_unpack('n')
+          return self.new(port)
```
Comment (olleolleolle) [inline lib/resolv.rb:1960]:
```suggestion
          new(port)
```

File: lib/resolv.rb [hunk 5]
```diff
@@ -1676,6 +1684,379 @@ def get_rr
...
+        KeyName = :'no-default-alpn'
+        KeyNumber = 2
+        ClassHash[KeyName] = ClassHash[KeyNumber] = self # :nodoc:
+
+        def encode(msg) # :nodoc:
+          # no payload
+        end
+
+        def self.decode(msg) # :nodoc:
+          return self.new
```
Comment (olleolleolle) [inline lib/resolv.rb:1930]:
```suggestion
          new
```

File: lib/resolv.rb [hunk 6]
```diff
@@ -1676,6 +1684,379 @@ def get_rr
...
+        def initialize(protocol_ids)
+          @protocol_ids = protocol_ids.map(&:to_str)
+        end
+
+        def encode(msg) # :nodoc:
+          msg.put_string_list(@protocol_ids)
+        end
+
+        def self.decode(msg) # :nodoc:
+          return self.new(msg.get_string_list)
```
Comment (olleolleolle) [inline lib/resolv.rb:1913]:
```suggestion
          new(msg.get_string_list)
```

File: lib/resolv.rb [hunk 7]
```diff
@@ -1676,6 +1684,379 @@ def get_rr
...
+
+        def encode(msg) # :nodoc:
+          @keys.sort.each do |key|
+            msg.put_pack('n', key)
+          end
+        end
+
+        def self.decode(msg) # :nodoc:
+          keys = msg.get_list { msg.get_unpack('n')[0] }
+          return self.new(keys)
```
Comment (olleolleolle) [inline lib/resolv.rb:1884]:
```suggestion
          new(keys)
```

File: lib/resolv.rb [hunk 8]
```diff
@@ -1676,6 +1684,379 @@ def get_rr
...
+        end
+
+        def self.create(key_number)
+          c = Class.new(Generic)
+          key_name = :"key#{key_number}"
+          c.const_set(:KeyName, key_name)
+          c.const_set(:KeyNumber, key_number)
+          self.const_set(:"Key#{key_number}", c)
+          ClassHash[key_name] = ClassHash[key_number] = c
+          return c
```
Comment (olleolleolle) [inline lib/resolv.rb:1852]:
```suggestion
          c
```

File: lib/resolv.rb [hunk 9]
```diff
@@ -1676,6 +1684,379 @@ def get_rr
...
+        def initialize(value)
+          @value = value
+        end
+
+        def encode(msg) # :nodoc:
+          msg.put_bytes(@value)
+        end
+
+        def self.decode(msg) # :nodoc:
+          return self.new(msg.get_bytes)
```
Comment (olleolleolle) [inline lib/resolv.rb:1842]:
```suggestion
          new(msg.get_bytes)
```

File: lib/resolv.rb [hunk 10]
```diff
@@ -1676,6 +1684,379 @@ def get_rr
...
+
+      def self.decode(msg) # :nodoc:
+        params = msg.get_list do
+          key, = msg.get_unpack('n')
+          msg.get_length16 do
+            SvcParam::ClassHash[key].decode(msg)
+          end
+        end
+
+        return self.new(params)
```
Comment (olleolleolle) [inline lib/resolv.rb:1768]:
```suggestion
        new(params)
```

**Review by olleolleolle — COMMENTED** [review]
Marked some of the usages of `self.new`. 

Perhaps these are an expression choice in this library. Perhaps I should offer a PR to change those wholesale.

**Comment by hanazuki** [comment]
The entire library seems to prefer the style that uses `self.new` and explicit `return` in multiple-sentence methods, which I wanted to preserve as much as possible to respect the original author. (I for myself have no preference for either style.)

**Comment by sorah** [comment]
> which I wanted to preserve as much as possible to respect the original author

This.

**Pull request merged**